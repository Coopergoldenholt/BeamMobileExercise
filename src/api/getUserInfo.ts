import axios from "axios"
import { IUser } from "../types"

export const getUserInfo = async (): Promise<IUser | string> => {
    try {
        let response = await axios.get('https://member-data.beam.dental/memberInfoHistory.json')

        let sortResp = response.data.sort((a: IUser, b: IUser) => {
            var effectiveDateA = a.effective_date
            var effectiveDateB = b.effective_date
            if (effectiveDateA < effectiveDateB) {
                return 1;
            }
            if (effectiveDateA > effectiveDateB) {
                return -1;
            }

            return 0;
        });

        return sortResp[0]
    }
    catch {
        return 'User not found'
    }
}