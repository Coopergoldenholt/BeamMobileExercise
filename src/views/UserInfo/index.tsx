import AsyncStorage from '@react-native-async-storage/async-storage'
import React, { useCallback, useEffect, useState } from 'react'
import { useIsFocused } from '@react-navigation/core'
import { View, Text, ActivityIndicator, ScrollView, RefreshControl, FlatList, TouchableOpacity, Alert, StyleSheet } from 'react-native'
import { getUserInfo } from '../../api/getUserInfo'
import { IBrushingRecord, IUser } from '../../types'
import BrushingRecordDisplay from './components/BrushingRecordDisplay'

interface UserInfoProps {
    navigation: any;
}

const wait = (timeout: number) => {
    return new Promise(resolve => setTimeout(resolve, timeout));
}

const UserInfo: React.FC<UserInfoProps> = ({ navigation }) => {
    const [userInfo, setUserInfo] = useState<IUser | null>(null)
    const [refreshing, setRefreshing] = useState(false);
    const [brushingRecords, setBrushingRecords] = useState<IBrushingRecord[]>([])
    const [invalidUser, setInvalidUser] = useState(false)
    const isFocused = useIsFocused()

    useEffect(() => {
        getUser()
    }, [])

    useEffect(() => {
        getData()
    }, [isFocused])

    const getData = async () => {
        try {
            const value = await AsyncStorage.getItem('brushingRecords')

            if (value !== null) {

                setBrushingRecords(JSON.parse(value))
            }
        } catch (e) {
            Alert.alert('Could not get your brushing record.')
        }
    }

    const alertBrushingRecord = (id: string) => {
        Alert.alert(
            "Delete Record?",
            "Are you sure you want to delete this brushing record?",
            [
                {
                    text: "Cancel",
                    style: "cancel"
                },
                { text: "OK", onPress: () => deleteBrushingRecord(id) }
            ]
        );
    }

    const deleteBrushingRecord = async (id: string) => {
        const filtRecords = brushingRecords.filter(rec => rec.id !== id)
        try {
            const stringifiedRecords = JSON.stringify(filtRecords)
            await AsyncStorage.setItem('brushingRecords', stringifiedRecords)
            setBrushingRecords(filtRecords)
        } catch (e) {
            Alert.alert('Could not delete your brushing record.')
        }
    }

    const onRefresh = useCallback(() => {
        setRefreshing(true);
        getUser()
        wait(2000).then(() => setRefreshing(false));
    }, []);

    const getUser = async () => {
        let info = await getUserInfo()
        if (typeof info === 'string') {
            setInvalidUser(true)
            setUserInfo(null)
        } else {
            setInvalidUser(false)
            //@ts-ignore
            setUserInfo(info)
        }
    }

    return (
        userInfo ?
            <View style={styles.container}>
                <ScrollView
                    refreshControl={
                        <RefreshControl
                            refreshing={refreshing}
                            onRefresh={onRefresh}
                        />
                    }
                >
                    <Text style={styles.nameText}>Welcome, {userInfo.name}</Text>
                    <View>
                        <Text style={styles.shippingText}>Shipping Address:</Text>
                        <Text style={styles.addressText}>
                            {userInfo.shipping_address}
                        </Text>
                        <Text style={styles.addressText}>
                            {userInfo.shipping_city}, {userInfo.shipping_state} {userInfo.shipping_zip_code}
                        </Text>
                    </View>
                </ScrollView>
                <TouchableOpacity
                    testID='CreateRecord.Button'
                    style={styles.button}
                    onPress={() => navigation.navigate('BrushingRecord')}
                >
                    <Text style={styles.buttonText}>Create Record</Text>
                </TouchableOpacity>

                {brushingRecords.length > 0 ?
                    <FlatList
                        data={brushingRecords}
                        renderItem={({ item }: { item: IBrushingRecord }) => <BrushingRecordDisplay
                            record={item}
                            alertBrushingRecord={alertBrushingRecord}
                            navigation={navigation}
                        />
                        }
                        keyExtractor={item => item.id}
                    /> :
                    <Text testID='no-records'>You don't have any brushing records to display.</Text>
                }
            </View>


            : invalidUser ? <Text>Could not find User. Please try again.</Text> : <ActivityIndicator />
    )
}

const styles = StyleSheet.create({
    container: {
        padding: 10
    },
    nameText: {
        fontSize: 22,
        paddingBottom: 15
    },
    shippingText: {
        fontSize: 20,
        paddingBottom: 8
    },
    addressText: {
        fontSize: 18
    },
    button: {
        alignSelf: 'center',
        backgroundColor: '#20a9ff',
        borderRadius: 5,
        height: 40,
        width: 200,
        alignItems: 'center',
        justifyContent: 'center',
        marginTop: 30,
        marginBottom: 20
    },
    buttonText: {
        color: 'white'
    },
})

export default UserInfo