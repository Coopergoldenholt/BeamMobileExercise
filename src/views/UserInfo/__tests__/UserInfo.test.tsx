import React from 'react'
import { render, fireEvent } from 'react-native-testing-library'

import UserInfo from '..'
import { getUserInfo } from '../../../api/getUserInfo'

jest.mock('@react-navigation/core');


const navigation = {
    navigate: jest.fn(),
    goBack: jest.fn()
};


it('renders default elements', async () => {
    const { toJSON } = render(<UserInfo navigation />)
    expect(toJSON()).toMatchSnapshot();
})

// it('create record moves to new screen', async () => {
//     const { getByText } = render(
//         <UserInfo navigation={navigation} />
//     )

//     fireEvent.press(getByText('Create Record'))

// })
