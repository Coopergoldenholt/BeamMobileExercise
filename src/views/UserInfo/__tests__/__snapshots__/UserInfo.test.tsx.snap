// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default elements 1`] = `<ActivityIndicator />`;
